File: frontend/src/api/pipelines.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiRequest } from "./client";

export type WorkflowStatus = "DRAFT" | "PUBLISHED" | "ARCHIVED";
export type NodeType =
  // v1 (Phase 3.2)
  | "NOOP"
  | "SOURCE_API"
  | "SQL_TRANSFORM"
  | "DEDUP"
  | "DQ_CHECK"
  | "LOAD_MASTER"
  | "NOTIFY"
  // v2 (Phase 5 generic + Phase 6 Wave 1)
  | "MAP_FIELDS"
  | "SQL_INLINE_TRANSFORM"
  | "SQL_ASSET_TRANSFORM"
  | "HTTP_TRANSFORM"
  | "FUNCTION_TRANSFORM"
  | "LOAD_TARGET"
  | "OCR_TRANSFORM"
  | "CRAWL_FETCH"
  | "STANDARDIZE"
  | "SOURCE_DATA"
  | "PUBLIC_API_FETCH"
  // Phase 7 Wave 1A — 외부 push / upload / DB 수집
  | "WEBHOOK_INGEST"
  | "FILE_UPLOAD_INGEST"
  | "DB_INCREMENTAL_FETCH"
  // Phase 7 Wave 1B — OCR/Crawler push 결과
  | "OCR_RESULT_INGEST"
  | "CRAWLER_RESULT_INGEST"
  | "CDC_EVENT_FETCH";

export interface NodeOut {
  node_id: number;
  node_key: string;
  node_type: NodeType;
  config_json: Record<string, unknown>;
  position_x: number;
  position_y: number;
}

export interface EdgeOut {
  edge_id: number;
  from_node_id: number;
  to_node_id: number;
  condition_expr: Record<string, unknown> | null;
}

export interface WorkflowOut {
  workflow_id: number;
  name: string;
  version: number;
  description: string | null;
  status: WorkflowStatus;
  created_by: number | null;
  created_at: string;
  updated_at: string;
  published_at: string | null;
  // Phase 3.2.7
  schedule_cron: string | null;
  schedule_enabled: boolean;
}

export interface WorkflowDetail extends WorkflowOut {
  nodes: NodeOut[];
  edges: EdgeOut[];
}

export type NodeRunStatus =
  | "PENDING"
  | "READY"
  | "RUNNING"
  | "SUCCESS"
  | "FAILED"
  | "SKIPPED"
  | "CANCELLED";

export type PipelineRunStatus =
  | "PENDING"
  | "RUNNING"
  | "ON_HOLD"
  | "SUCCESS"
  | "FAILED"
  | "CANCELLED";

export interface NodeRunOut {
  node_run_id: number;
  node_definition_id: number;
  node_key: string;
  node_type: string;
  status: NodeRunStatus;
  attempt_no: number;
  started_at: string | null;
  finished_at: string | null;
  error_message: string | null;
  output_json: Record<string, unknown> | null;
}

export interface PipelineRunOut {
  pipeline_run_id: number;
  workflow_id: number;
  run_date: string;
  status: PipelineRunStatus;
  triggered_by: number | null;
  started_at: string | null;
  finished_at: string | null;
  error_message: string | null;
  created_at: string;
}

export interface PipelineRunDetail extends PipelineRunOut {
  node_runs: NodeRunOut[];
}

export function useWorkflows(params: { status?: WorkflowStatus; limit?: number } = {}) {
  return useQuery({
    queryKey: ["pipelines", "workflows", params],
    queryFn: () =>
      apiRequest<WorkflowOut[]>("/v1/pipelines", { params: { ...params } }),
  });
}

export function useWorkflowDetail(workflowId: number | null) {
  return useQuery({
    queryKey: ["pipelines", "workflows", workflowId],
    enabled: workflowId != null,
    queryFn: () =>
      apiRequest<WorkflowDetail>(`/v1/pipelines/${workflowId}`),
  });
}

export function usePipelineRun(pipelineRunId: number | null) {
  return useQuery({
    queryKey: ["pipelines", "runs", pipelineRunId],
    enabled: pipelineRunId != null,
    queryFn: () =>
      apiRequest<PipelineRunDetail>(`/v1/pipelines/runs/${pipelineRunId}`),
    refetchInterval: 5_000, // SSE 가 끊겨도 5s 폴링으로 fallback.
  });
}

// ---------------------------------------------------------------------------
// Designer mutations (Phase 3.2.4)
// ---------------------------------------------------------------------------
export interface NodeIn {
  node_key: string;
  node_type: NodeType;
  config_json?: Record<string, unknown>;
  position_x?: number;
  position_y?: number;
}

export interface EdgeIn {
  from_node_key: string;
  to_node_key: string;
  condition_expr?: Record<string, unknown> | null;
}

export interface WorkflowCreate {
  name: string;
  version?: number;
  description?: string | null;
  nodes: NodeIn[];
  edges?: EdgeIn[];
}

export interface WorkflowPatch {
  name?: string;
  description?: string | null;
  nodes?: NodeIn[];
  edges?: EdgeIn[];
}

export function useCreateWorkflow() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: WorkflowCreate) =>
      apiRequest<WorkflowDetail>("/v1/pipelines", {
        method: "POST",
        body,
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["pipelines", "workflows"] });
    },
  });
}

export function useUpdateWorkflow() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      workflowId,
      body,
    }: {
      workflowId: number;
      body: WorkflowPatch;
    }) =>
      apiRequest<WorkflowDetail>(`/v1/pipelines/${workflowId}`, {
        method: "PATCH",
        body,
      }),
    onSuccess: (_, vars) => {
      qc.invalidateQueries({
        queryKey: ["pipelines", "workflows", vars.workflowId],
      });
      qc.invalidateQueries({ queryKey: ["pipelines", "workflows"] });
    },
  });
}

// Phase 3.2.6: 응답이 단순 WorkflowOut 이 아니라 transition 메타로 확장됨.
export interface PipelineReleaseOut {
  release_id: number;
  workflow_name: string;
  version_no: number;
  source_workflow_id: number | null;
  released_workflow_id: number;
  released_by: number | null;
  released_at: string;
  change_summary: {
    added?: string[];
    removed?: string[];
    changed?: string[];
    edges_added?: string[];
    edges_removed?: string[];
  };
}

export interface PipelineReleaseDetail extends PipelineReleaseOut {
  nodes_snapshot: Array<Record<string, unknown>>;
  edges_snapshot: Array<Record<string, unknown>>;
}

export interface WorkflowStatusTransitionOut {
  workflow: WorkflowOut;
  published_workflow: WorkflowOut | null;
  release: PipelineReleaseOut | null;
}

export function useTransitionWorkflowStatus() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      workflowId,
      status,
    }: {
      workflowId: number;
      status: "PUBLISHED" | "ARCHIVED";
    }) =>
      apiRequest<WorkflowStatusTransitionOut>(
        `/v1/pipelines/${workflowId}/status`,
        { method: "PATCH", body: { status } },
      ),
    onSuccess: (_, vars) => {
      qc.invalidateQueries({
        queryKey: ["pipelines", "workflows", vars.workflowId],
      });
      qc.invalidateQueries({ queryKey: ["pipelines", "workflows"] });
      qc.invalidateQueries({ queryKey: ["pipelines", "releases"] });
    },
  });
}

export function usePipelineReleases(name?: string | null) {
  return useQuery({
    queryKey: ["pipelines", "releases", name ?? null],
    queryFn: () =>
      apiRequest<PipelineReleaseOut[]>("/v1/pipelines/releases", {
        params: { name: name ?? undefined, limit: 200 },
      }),
  });
}

export function usePipelineReleaseDetail(releaseId: number | null) {
  return useQuery({
    queryKey: ["pipelines", "releases", "detail", releaseId],
    enabled: releaseId != null,
    queryFn: () =>
      apiRequest<PipelineReleaseDetail>(`/v1/pipelines/releases/${releaseId}`),
  });
}

export interface NodeChangeOut {
  node_key: string;
  node_type: string | null;
  config_before: Record<string, unknown> | null;
  config_after: Record<string, unknown> | null;
}
export interface EdgeChangeOut {
  from_node_key: string;
  to_node_key: string;
}
export interface WorkflowDiffOut {
  before_workflow_id: number;
  after_workflow_id: number;
  nodes_added: NodeChangeOut[];
  nodes_removed: NodeChangeOut[];
  nodes_changed: NodeChangeOut[];
  edges_added: EdgeChangeOut[];
  edges_removed: EdgeChangeOut[];
}

export function useWorkflowDiff(
  workflowId: number | null,
  againstId: number | null,
) {
  return useQuery({
    queryKey: ["pipelines", "diff", workflowId, againstId],
    enabled: workflowId != null && againstId != null,
    queryFn: () =>
      apiRequest<WorkflowDiffOut>(`/v1/pipelines/${workflowId}/diff`, {
        params: { against: againstId! },
      }),
  });
}

export function useTriggerRun() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (workflowId: number) =>
      apiRequest<PipelineRunDetail>(`/v1/pipelines/${workflowId}/runs`, {
        method: "POST",
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["pipelines", "runs"] });
    },
  });
}

// ---------------------------------------------------------------------------
// Schedule / Backfill / Restart / Search (Phase 3.2.7)
// ---------------------------------------------------------------------------
export function useUpdateSchedule() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      workflowId,
      cron,
      enabled,
    }: {
      workflowId: number;
      cron: string | null;
      enabled: boolean;
    }) =>
      apiRequest<WorkflowOut>(`/v1/pipelines/${workflowId}/schedule`, {
        method: "PATCH",
        body: { cron, enabled },
      }),
    onSuccess: (_, vars) => {
      qc.invalidateQueries({
        queryKey: ["pipelines", "workflows", vars.workflowId],
      });
      qc.invalidateQueries({ queryKey: ["pipelines", "workflows"] });
    },
  });
}

export interface BackfillResponse {
  pipeline_run_ids: number[];
  run_dates: string[];
}

export function useBackfill() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      workflowId,
      start_date,
      end_date,
    }: {
      workflowId: number;
      start_date: string;
      end_date: string;
    }) =>
      apiRequest<BackfillResponse>(`/v1/pipelines/${workflowId}/backfill`, {
        method: "POST",
        body: { start_date, end_date },
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["pipelines", "runs"] });
    },
  });
}

export interface RunSearchParams {
  workflow_id?: number | null;
  status?: PipelineRunStatus | null;
  from?: string | null; // YYYY-MM-DD
  to?: string | null;
  limit?: number;
  offset?: number;
}

export function useSearchRuns(params: RunSearchParams = {}) {
  return useQuery({
    queryKey: ["pipelines", "runs", "search", params],
    queryFn: () =>
      apiRequest<PipelineRunOut[]>("/v1/pipelines/runs", {
        params: {
          workflow_id: params.workflow_id ?? undefined,
          status: params.status ?? undefined,
          from: params.from ?? undefined,
          to: params.to ?? undefined,
          limit: params.limit ?? 100,
          offset: params.offset ?? 0,
        },
      }),
  });
}

export interface RestartResponse {
  new_pipeline_run_id: number;
  new_run_date: string;
  ready_node_run_ids: number[];
  seeded_success_node_keys: string[];
}

export function useRestartRun() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      runId,
      from_node_key,
    }: {
      runId: number;
      from_node_key?: string | null;
    }) =>
      apiRequest<RestartResponse>(`/v1/pipelines/runs/${runId}/restart`, {
        method: "POST",
        body: { from_node_key: from_node_key ?? null },
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["pipelines", "runs"] });
    },
  });
}

// ---------------------------------------------------------------------------
// DQ Gate (Phase 4.2.2) — ON_HOLD list + approve/reject.
// ---------------------------------------------------------------------------
export interface QualityResultOut {
  quality_result_id: number;
  pipeline_run_id: number | null;
  node_run_id: number | null;
  target_table: string;
  check_kind: string;
  passed: boolean;
  severity: string;
  status: "PASS" | "WARN" | "FAIL";
  details_json: Record<string, unknown>;
  sample_json: Array<Record<string, unknown>>;
  created_at: string;
}

export interface OnHoldRunOut extends PipelineRunOut {
  failed_node_keys: string[];
  quality_results: QualityResultOut[];
}

export interface HoldDecisionResponse {
  decision_id: number;
  pipeline_run_id: number;
  run_date: string;
  decision: "APPROVE" | "REJECT";
  pipeline_status: string;
  ready_node_run_ids: number[];
  cancelled_node_run_ids: number[];
  rollback_rows: number;
}

export function useOnHoldRuns(params: { limit?: number; offset?: number } = {}) {
  return useQuery({
    queryKey: ["pipelines", "runs", "on_hold", params],
    queryFn: () =>
      apiRequest<OnHoldRunOut[]>("/v1/pipelines/runs/on_hold", {
        params: { limit: params.limit ?? 50, offset: params.offset ?? 0 },
      }),
    refetchInterval: 10_000,
  });
}

export function useApproveHold() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      runId,
      reason,
    }: {
      runId: number;
      reason?: string | null;
    }) =>
      apiRequest<HoldDecisionResponse>(
        `/v1/pipelines/runs/${runId}/hold/approve`,
        { method: "POST", body: { reason: reason ?? null } },
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["pipelines", "runs"] });
    },
  });
}

export function useRejectHold() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      runId,
      reason,
    }: {
      runId: number;
      reason?: string | null;
    }) =>
      apiRequest<HoldDecisionResponse>(
        `/v1/pipelines/runs/${runId}/hold/reject`,
        { method: "POST", body: { reason: reason ?? null } },
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["pipelines", "runs"] });
    },
  });
}

File: backend/app/schemas/pipelines.py
"""Pipeline (Visual ETL) API schemas (Phase 3.2.1)."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

NodeType = Literal[
    # v1 (Phase 3.2)
    "NOOP",
    "SOURCE_API",
    "SQL_TRANSFORM",
    "DEDUP",
    "DQ_CHECK",
    "LOAD_MASTER",
    "NOTIFY",
    # v2 (Phase 5 generic + Phase 6 Wave 1) — DB CHECK (migration 0047) 와 정렬
    "MAP_FIELDS",
    "SQL_INLINE_TRANSFORM",
    "SQL_ASSET_TRANSFORM",
    "HTTP_TRANSFORM",
    "FUNCTION_TRANSFORM",
    "LOAD_TARGET",
    "OCR_TRANSFORM",
    "CRAWL_FETCH",
    "STANDARDIZE",
    "SOURCE_DATA",
    "PUBLIC_API_FETCH",
    # Phase 7 Wave 1A — 외부 push / upload / DB 수집
    "WEBHOOK_INGEST",
    "FILE_UPLOAD_INGEST",
    "DB_INCREMENTAL_FETCH",
    # Phase 7 Wave 1B — OCR/Crawler push 결과
    "OCR_RESULT_INGEST",
    "CRAWLER_RESULT_INGEST",
    "CDC_EVENT_FETCH",
]
WorkflowStatus = Literal["DRAFT", "PUBLISHED", "ARCHIVED"]


# ---------------------------------------------------------------------------
# Definitions (input)
# ---------------------------------------------------------------------------
class NodeIn(BaseModel):
    node_key: str = Field(min_length=1, max_length=64)
    node_type: NodeType
    config_json: dict[str, Any] = Field(default_factory=dict)
    position_x: int = 0
    position_y: int = 0


class EdgeIn(BaseModel):
    from_node_key: str = Field(min_length=1, max_length=64)
    to_node_key: str = Field(min_length=1, max_length=64)
    condition_expr: dict[str, Any] | None = None


class WorkflowCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    version: int = Field(default=1, ge=1)
    description: str | None = Field(default=None, max_length=1000)
    nodes: list[NodeIn]
    edges: list[EdgeIn] = Field(default_factory=list)


class WorkflowPatch(BaseModel):
    """DRAFT 상태에서만 수정 — name/description/nodes/edges 교체 가능. status 전이는 별도."""

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=1000)
    nodes: list[NodeIn] | None = None
    edges: list[EdgeIn] | None = None


class WorkflowStatusUpdate(BaseModel):
    status: Literal["PUBLISHED", "ARCHIVED"]


# ---------------------------------------------------------------------------
# Output
# ---------------------------------------------------------------------------
class NodeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    node_id: int
    node_key: str
    node_type: NodeType
    config_json: dict[str, Any]
    position_x: int
    position_y: int


class EdgeOut(BaseModel):
    edge_id: int
    from_node_id: int
    to_node_id: int
    condition_expr: dict[str, Any] | None


class WorkflowOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    workflow_id: int
    name: str
    version: int
    description: str | None
    status: WorkflowStatus
    created_by: int | None
    created_at: datetime
    updated_at: datetime
    published_at: datetime | None
    # Phase 3.2.7
    schedule_cron: str | None = None
    schedule_enabled: bool = False


class WorkflowDetail(WorkflowOut):
    nodes: list[NodeOut] = Field(default_factory=list)
    edges: list[EdgeOut] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Run
# ---------------------------------------------------------------------------
class NodeRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    node_run_id: int
    node_definition_id: int
    node_key: str
    node_type: str
    status: str
    attempt_no: int
    started_at: datetime | None
    finished_at: datetime | None
    error_message: str | None
    output_json: dict[str, Any] | None


class PipelineRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    pipeline_run_id: int
    workflow_id: int
    run_date: date
    status: str
    triggered_by: int | None
    started_at: datetime | None
    finished_at: datetime | None
    error_message: str | None
    created_at: datetime


class PipelineRunDetail(PipelineRunOut):
    node_runs: list[NodeRunOut] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Release / Diff (Phase 3.2.6)
# ---------------------------------------------------------------------------
class PipelineReleaseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    release_id: int
    workflow_name: str
    version_no: int
    source_workflow_id: int | None
    released_workflow_id: int
    released_by: int | None
    released_at: datetime
    change_summary: dict[str, Any]


class PipelineReleaseDetail(PipelineReleaseOut):
    nodes_snapshot: list[dict[str, Any]] = Field(default_factory=list)
    edges_snapshot: list[dict[str, Any]] = Field(default_factory=list)


class WorkflowStatusTransitionOut(BaseModel):
    """PATCH /status 의 응답 — PUBLISHED 시 새 워크플로/release 정보를 함께 반환."""

    workflow: WorkflowOut
    published_workflow: WorkflowOut | None = None
    release: PipelineReleaseOut | None = None


class NodeChangeOut(BaseModel):
    node_key: str
    node_type: str | None = None
    config_before: dict[str, Any] | None = None
    config_after: dict[str, Any] | None = None


class EdgeChangeOut(BaseModel):
    from_node_key: str
    to_node_key: str


class WorkflowDiffOut(BaseModel):
    before_workflow_id: int
    after_workflow_id: int
    nodes_added: list[NodeChangeOut] = Field(default_factory=list)
    nodes_removed: list[NodeChangeOut] = Field(default_factory=list)
    nodes_changed: list[NodeChangeOut] = Field(default_factory=list)
    edges_added: list[EdgeChangeOut] = Field(default_factory=list)
    edges_removed: list[EdgeChangeOut] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Schedule / Backfill / Restart (Phase 3.2.7)
# ---------------------------------------------------------------------------
class ScheduleUpdate(BaseModel):
    cron: str | None = Field(default=None, max_length=200)
    enabled: bool = False


class BackfillRequest(BaseModel):
    """`[start_date, end_date]` 양 끝 포함."""

    start_date: date
    end_date: date


class BackfillResponse(BaseModel):
    pipeline_run_ids: list[int]
    run_dates: list[date]


class RestartRequest(BaseModel):
    from_node_key: str | None = Field(default=None, min_length=1, max_length=64)


class RestartResponse(BaseModel):
    new_pipeline_run_id: int
    new_run_date: date
    ready_node_run_ids: list[int]
    seeded_success_node_keys: list[str]


# ---------------------------------------------------------------------------
# DQ Gate (Phase 4.2.2)
# ---------------------------------------------------------------------------
class HoldDecisionRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=2000)


class QualityResultOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    quality_result_id: int
    pipeline_run_id: int | None
    node_run_id: int | None
    target_table: str
    check_kind: str
    passed: bool
    severity: str
    status: str
    details_json: dict[str, Any]
    sample_json: list[dict[str, Any]]
    created_at: datetime


class HoldDecisionResponse(BaseModel):
    decision_id: int
    pipeline_run_id: int
    run_date: date
    decision: Literal["APPROVE", "REJECT"]
    pipeline_status: str
    ready_node_run_ids: list[int] = Field(default_factory=list)
    cancelled_node_run_ids: list[int] = Field(default_factory=list)
    rollback_rows: int = 0


class OnHoldRunOut(PipelineRunOut):
    """ON_HOLD pipeline_run 1건 + 실패 DQ 결과 미리보기."""

    failed_node_keys: list[str] = Field(default_factory=list)
    quality_results: list[QualityResultOut] = Field(default_factory=list)


__all__ = [
    "BackfillRequest",
    "BackfillResponse",
    "EdgeChangeOut",
    "EdgeIn",
    "EdgeOut",
    "HoldDecisionRequest",
    "HoldDecisionResponse",
    "NodeChangeOut",
    "NodeIn",
    "NodeOut",
    "NodeRunOut",
    "NodeType",
    "OnHoldRunOut",
    "PipelineReleaseDetail",
    "PipelineReleaseOut",
    "PipelineRunDetail",
    "PipelineRunOut",
    "QualityResultOut",
    "RestartRequest",
    "RestartResponse",
    "ScheduleUpdate",
    "WorkflowCreate",
    "WorkflowDetail",
    "WorkflowDiffOut",
    "WorkflowOut",
    "WorkflowPatch",
    "WorkflowStatus",
    "WorkflowStatusTransitionOut",
    "WorkflowStatusUpdate",
]
